File: studio-web/src/i18n/registry.ts
import meta from "./locale-meta.json";

export type LocaleDirection = "ltr" | "rtl";
export type LanguagePackStatus =
  | "DRAFT"
  | "IN_REVIEW"
  | "VERIFIED"
  | "ACTIVE"
  | "DEPRECATED";
export type LinguisticReviewLevel =
  | "KEY_PRESENT"
  | "MACHINE_DRAFT"
  | "LINGUISTICALLY_REVIEWED"
  | "PRODUCTION_APPROVED"
  | "HUMAN_APPROVED";

export interface LocaleDescriptor {
  locale: string;
  displayName: string;
  nativeName: string;
  direction: LocaleDirection;
  fallback: string;
  enabled: boolean;
  status: LanguagePackStatus;
  reviewLevel: LinguisticReviewLevel;
  dateLocale: string;
  numberLocale: string;
  notes?: string;
}

export const NAMESPACES = meta.namespaces as readonly string[];
export const BETA_CRITICAL_NAMESPACES = meta.betaCriticalNamespaces as readonly string[];
export const LOCALE_REGISTRY: readonly LocaleDescriptor[] = meta.locales as LocaleDescriptor[];

export const DEFAULT_LOCALE = "en";
export const SUPPORTED_LOCALES = LOCALE_REGISTRY.map((l) => l.locale);

export function getLocale(locale: string): LocaleDescriptor | undefined {
  return LOCALE_REGISTRY.find((l) => l.locale === locale);
}

export function isSupportedLocale(locale: string): boolean {
  return SUPPORTED_LOCALES.includes(locale);
}

export function activeLocales(): LocaleDescriptor[] {
  return LOCALE_REGISTRY.filter((l) => l.enabled && l.status === "ACTIVE");
}

export function resolveSupportedLocale(candidate: string | null | undefined): string | null {
  if (!candidate) return null;
  const raw = candidate.trim();
  if (!raw) return null;
  if (isSupportedLocale(raw)) return raw;
  const lower = raw.toLowerCase();
  if (lower.startsWith("zh")) {
    if (lower.includes("hant") || lower.includes("tw") || lower.includes("hk")) return null;
    if (isSupportedLocale("zh-Hans")) return "zh-Hans";
  }
  const primary = lower.split("-")[0];
  if (primary === "pt" && isSupportedLocale("pt")) return "pt";
  if (isSupportedLocale(primary)) return primary;
  return null;
}

File: studio-api/app/codirector/language_intelligence/registry.py
"""Canonical locale registry (mirrors studio-web/src/i18n)."""

from __future__ import annotations

from typing import Any, Optional

LOCALE_REGISTRY: list[dict[str, Any]] = [
    {
        "locale": "en",
        "displayName": "English",
        "nativeName": "English",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "HUMAN_APPROVED",
        "dateLocale": "en",
        "numberLocale": "en",
    },
    {
        "locale": "fr",
        "displayName": "French",
        "nativeName": "Français",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "fr",
        "numberLocale": "fr",
    },
    {
        "locale": "es",
        "displayName": "Spanish",
        "nativeName": "Español",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "es",
        "numberLocale": "es",
    },
    {
        "locale": "ja",
        "displayName": "Japanese",
        "nativeName": "日本語",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "ja",
        "numberLocale": "ja",
    },
    {
        "locale": "zh-Hans",
        "displayName": "Chinese (Simplified)",
        "nativeName": "简体中文",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "zh-CN",
        "numberLocale": "zh-CN",
    },
    {
        "locale": "hi",
        "displayName": "Hindi",
        "nativeName": "हिन्दी",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "hi",
        "numberLocale": "hi",
    },
    {
        "locale": "ru",
        "displayName": "Russian",
        "nativeName": "Русский",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "ru",
        "numberLocale": "ru",
    },
    {
        "locale": "pt",
        "displayName": "Portuguese",
        "nativeName": "Português",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "pt-BR",
        "numberLocale": "pt-BR",
        "notes": "Brazilian Portuguese baseline for locale id pt",
    },
    {
        "locale": "ar",
        "displayName": "Arabic",
        "nativeName": "العربية",
        "direction": "rtl",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "ar",
        "numberLocale": "ar",
    },
    {
        "locale": "bn",
        "displayName": "Bengali",
        "nativeName": "বাংলা",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "bn",
        "numberLocale": "bn",
    },
    {
        "locale": "id",
        "displayName": "Indonesian",
        "nativeName": "Bahasa Indonesia",
        "direction": "ltr",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "id",
        "numberLocale": "id",
    },
    {
        "locale": "ur",
        "displayName": "Urdu",
        "nativeName": "اردو",
        "direction": "rtl",
        "fallback": "en",
        "enabled": True,
        "status": "ACTIVE",
        "reviewLevel": "MACHINE_DRAFT",
        "dateLocale": "ur",
        "numberLocale": "ur",
    },
]

SUPPORTED = {row["locale"] for row in LOCALE_REGISTRY}


def get_locale(locale: str) -> Optional[dict[str, Any]]:
    for row in LOCALE_REGISTRY:
        if row["locale"] == locale:
            return row
    return None


def is_supported_locale(locale: str) -> bool:
    return locale in SUPPORTED


def resolve_locale(candidate: Optional[str]) -> str:
    if not candidate:
        return "en"
    if candidate in SUPPORTED:
        return candidate
    lower = candidate.lower()
    if lower.startswith("zh") and "zh-Hans" in SUPPORTED:
        if "hant" in lower or "tw" in lower or "hk" in lower:
            return "en"
        return "zh-Hans"
    primary = lower.split("-")[0]
    if primary in SUPPORTED:
        return primary
    return "en"
